### ISSUE TITLE
The denoised image of the test is displayed in full black

### ISSUE THREAD

**Issue by ouhaochen9** [junbopeng/CBCT2CT_cDDPM#1]
May I ask whether the input CT and CBCT are denoised and de-denoised according to the one-dimensional gray image input into the Dataloader?
After noise is predicted using GaussianDiffusionTrainer in Diffusion_condition.py provided by the author, GaussianDiffusionSampler is used to remove noise from the verification set. The result picture is shown as all black. Does the author have this problem?

**Comment by ALIKAOoO** [comment]
请问from Diffusion import GaussianDiffusionTrainer_cond中的Diffusion是作者自定义的文件嘛？

**Comment by junbopeng** [comment]
> May I ask whether the input CT and CBCT are denoised and de-denoised according to the one-dimensional gray image input into the Dataloader? After noise is predicted using GaussianDiffusionTrainer in Diffusion_condition.py provided by the author, GaussianDiffusionSampler is used to remove noise from the verification set. The result picture is shown as all black. Does the author have this problem?

Sorry for the late response.
- For the model training, input CT and CBCT (into Dataloader) are both clean images, and the CT image will be perturbed by Gaussian noise in the trainer function.
- I have not encountered the "all black" problem. Maybe you can provide more details about it.

**Comment by junbopeng** [comment]
> 请问from Diffusion import GaussianDiffusionTrainer_cond中的Diffusion是作者自定义的文件嘛？

Hi,

The module should be imported from "Diffusion_condition.py" and the script of train_condition.py has been updated. (Line 16).

**Comment by QijingGJ** [comment]
Hello, may I ask if the input of ct is random noise during the test? Why the model is not stable during my test, sometimes it can generate normal images, and sometimes it generates all noise.

Thank you very much for your help.

**Comment by junbopeng** [comment]
> Hello, may I ask if the input of ct is random noise during the test? Why the model is not stable during my test, sometimes it can generate normal images, and sometimes it generates all noise.
> 
> Thank you very much for your help.

Yes, the initial input of CT is a standard Gaussian noise during the inference stage. Could you provide more details about "sometimes it generates all noise"?

**Comment by ALIKAOoO** [comment]
> > 请问from Diffusion import GaussianDiffusionTrainer_cond中的Diffusion是作者自定义的文件嘛？
> 
> 你好，
> 
> 该模块应从“Diffusion_condition.py”导入，并且 train_condition.py 的脚本已更新。 （第 16 行）。

Sorry,I mean that the introduced GaussianDiffusionTrainer_cond is not declared in the Diffusion_condition.py file.
GaussianDiffusionTrainer_cond==GaussianDiffusionTrainer?
please

**Comment by junbopeng** [comment]
> > > 请问from Diffusion import GaussianDiffusionTrainer_cond中的Diffusion是作者自定义的文件嘛？
> > 
> > 
> > 你好，
> > 该模块应从“Diffusion_condition.py”导入，并且 train_condition.py 的脚本已更新。 （第 16 行）。
> 
> Sorry,I mean that the introduced GaussianDiffusionTrainer_cond is not declared in the Diffusion_condition.py file. GaussianDiffusionTrainer_cond==GaussianDiffusionTrainer? please

Sorry, the class name was wrong and it has been corrected

**Comment by skyknights** [comment]
Hello, I trained and saved the model through GaussianDiffusionTrainer, but my training set is in png format instead of npy format, but I converted it. For the test code GaussianDiffusionSampler, how should I write it for testing? Are the test inputs still cbct and ct?

**Comment by QijingGJ** [comment]
> > Hello, may I ask if the input of ct is random noise during the test? Why the model is not stable during my test, sometimes it can generate normal images, and sometimes it generates all noise.
> > Thank you very much for your help.
> 
> Yes, the initial input of CT is a standard Gaussian noise during the inference stage. Could you provide more details about "sometimes it generates all noise"?

Hello, may I ask whether it is reasonable to return torch.clip(x_0, -1,1) in GaussianDiffusionSampler during the test? x_0 contains cbct and generated ct. The range of cbct is [-1,1], but the range of generated ct in my test is all less than -1. As a result, the generated image is all black, I modified the code as follows, but the generated image is not stable, most of the time is noise, and occasionally it can be faintly seen.

x_0 = x_0[:, 0, :, :]
min_val = x_0.min()
max_val = x_0.max()
normalized_tensor = (x_0 - min_val) / (max_val - min_val) * 2 - 1

Thank you very much for your help

**Comment by skyknights** [comment]
> > > Hello, may I ask if the input of ct is random noise during the test? Why the model is not stable during my test, sometimes it can generate normal images, and sometimes it generates all noise.
> > > Thank you very much for your help.
> > 
> > 
> > Yes, the initial input of CT is a standard Gaussian noise during the inference stage. Could you provide more details about "sometimes it generates all noise"?
> 
> Hello, may I ask whether it is reasonable to return torch.clip(x_0, -1,1) in GaussianDiffusionSampler during the test? x_0 contains cbct and generated ct. The range of cbct is [-1,1], but the range of generated ct in my test is all less than -1. As a result, the generated image is all black, I modified the code as follows, but the generated image is not stable, most of the time is noise, and occasionally it can be faintly seen.
> 
> x_0 = x_0[:, 0, :, :] min_val = x_0.min() max_val = x_0.max() normalized_tensor = (x_0 - min_val) / (max_val - min_val) * 2 - 1
> 
> Thank you very much for your help 你好，请问在测试期间返回GaussianDiffusionSampler中的torch . clip（x _ 0，-1，1）是否合理？x_0包含cbct和生成的ct。cbct的范围是【-1，1】，但是在我的测试中生成的ct的范围都小于-1。结果生成的图像全黑，我修改了如下代码，但生成的图像并不稳定，大部分时间都是噪点，偶尔还能隐约看到。x_0 = x_0【:，0，:】min _ val = x _ 0 . min（）max _ val = x _ 0 . max（）normalized _ tensor =（x _ 0-min _ val）/（max _ val-min _ val）* 2-1非常感谢您的帮助

我不清楚作者的代码部分是否完全正确 我在GaussianDiffusionSampler也出现了一些错误，可以加q1121269990讨论下

**Comment by HIT-wwb** [comment]
@QijingGJ Hello, I'm currently reproducing this work, but I'm having trouble writing test code, I don't know if I can communicate with you, here's my email: 18394961357@163.com or if you have QQ, this is my QQ: 1628556295

**Comment by ZhaKiNg** [comment]
> > > > Hello, may I ask if the input of ct is random noise during the test? Why the model is not stable during my test, sometimes it can generate normal images, and sometimes it generates all noise.
> > > > Thank you very much for your help.
> > > 
> > > 
> > > Yes, the initial input of CT is a standard Gaussian noise during the inference stage. Could you provide more details about "sometimes it generates all noise"?
> > 
> > 
> > Hello, may I ask whether it is reasonable to return torch.clip(x_0, -1,1) in GaussianDiffusionSampler during the test? x_0 contains cbct and generated ct. The range of cbct is [-1,1], but the range of generated ct in my test is all less than -1. As a result, the generated image is all black, I modified the code as follows, but the generated image is not stable, most of the time is noise, and occasionally it can be faintly seen.
> > x_0 = x_0[:, 0, :, :] min_val = x_0.min() max_val = x_0.max() normalized_tensor = (x_0 - min_val) / (max_val - min_val) * 2 - 1
> > Thank you very much for your help 你好，请问在测试期间返回GaussianDiffusionSampler中的torch . clip（x _ 0，-1，1）是否合理？x_0包含cbct和生成的ct。cbct的范围是【-1，1】，但是在我的测试中生成的ct的范围都小于-1。结果生成的图像全黑，我修改了如下代码，但生成的图像并不稳定，大部分时间都是噪点，偶尔还能隐约看到。x_0 = x_0【:，0，:】min _ val = x _ 0 . min（）max _ val = x _ 0 . max（）normalized _ tensor =（x _ 0-min _ val）/（max _ val-min _ val）* 2-1非常感谢您的帮助
> 
> 我不清楚作者的代码部分是否完全正确 我在GaussianDiffusionSampler也出现了一些错误，可以加q1121269990讨论下

兄弟 你写了test的部分吗 可以交流一下吗 有偿

**Comment by junbopeng** [comment]
> > > > > Hello, may I ask if the input of ct is random noise during the test? Why the model is not stable during my test, sometimes it can generate normal images, and sometimes it generates all noise.
> > > > > Thank you very much for your help.
> > > > 
> > > > 
> > > > Yes, the initial input of CT is a standard Gaussian noise during the inference stage. Could you provide more details about "sometimes it generates all noise"?
> > > 
> > > 
> > > Hello, may I ask whether it is reasonable to return torch.clip(x_0, -1,1) in GaussianDiffusionSampler during the test? x_0 contains cbct and generated ct. The range of cbct is [-1,1], but the range of generated ct in my test is all less than -1. As a result, the generated image is all black, I modified the code as follows, but the generated image is not stable, most of the time is noise, and occasionally it can be faintly seen.
> > > x_0 = x_0[:, 0, :, :] min_val = x_0.min() max_val = x_0.max() normalized_tensor = (x_0 - min_val) / (max_val - min_val) * 2 - 1
> > > Thank you very much for your help 你好，请问在测试期间返回GaussianDiffusionSampler中的torch . clip（x _ 0，-1，1）是否合理？x_0包含cbct和生成的ct。cbct的范围是【-1，1】，但是在我的测试中生成的ct的范围都小于-1。结果生成的图像全黑，我修改了如下代码，但生成的图像并不稳定，大部分时间都是噪点，偶尔还能隐约看到。x_0 = x_0【:，0，:】min _ val = x _ 0 . min（）max _ val = x _ 0 . max（）normalized _ tensor =（x _ 0-min _ val）/（max _ val-min _ val）* 2-1非常感谢您的帮助
> > 
> > 
> > 我不清楚作者的代码部分是否完全正确 我在GaussianDiffusionSampler也出现了一些错误，可以加q1121269990讨论下
> 
> 兄弟 你写了test的部分吗 可以交流一下吗 有偿

The script of test_condition.py has been uploaded.
